efi/image.go

```diff
@@ -0,0 +1,240 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package efi
+
+import (
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/snapcore/secboot"
+	"github.com/snapcore/snapd/snap"
+)
+
+// ImageReader corresponds to an open handle from which to read a binary image from.
+type ImageReader interface {
+	io.ReaderAt
+	io.Closer
+	Size() int64
+}
+
+// Image provides an image that is loaded during the boot process before ExitBootServices.
+type Image interface {
+	fmt.Stringer
+	Open() (ImageReader, error) // Open a handle to the image for reading
+}
+
+// SnapFileImage provides an image contained within a snap package that is loaded
+// during the boot process before ExitBootServices.
+type SnapFileImage struct {
+	Container snap.Container
+	FileName  string // The filename within the snap squashfs
+}
+
+// NewSnapFileImage creates a new SnapFileImage for the file at the
+// specified relative path within the supplied snap.
+func NewSnapFileImage(container snap.Container, filename string) *SnapFileImage {
+	return &SnapFileImage{
+		Container: container,
+		FileName:  filename}
+}
+
+// String implements [fmt.Stringer].
+func (f SnapFileImage) String() string {
+	return fmt.Sprintf("%#v:%s", f.Container, f.FileName)
+}
+
+// Open implements [ImageSource.Open].
+func (f SnapFileImage) Open() (ImageReader, error) {
+	return f.Container.RandomAccessFile(f.FileName)
+}
+
+type fileImageReader struct {
+	*os.File
+	size int64
+}
+
+func (h *fileImageReader) Size() int64 {
+	return h.size
+}
+
+// FileImage provides an image from a file that is loaded during the boot process
+// before ExitBootServices.
+type FileImage string
+
+// NewFileImage creates a new FileImage for the file at the specified path.
+func NewFileImage(path string) FileImage {
+	return FileImage(path)
+}
+
+// String implements [fmt.Stringer].
+func (p FileImage) String() string {
+	return string(p)
+}
+
+// Open implements [ImageSource.Open].
+func (p FileImage) Open() (ImageReader, error) {
+	f, err := os.Open(string(p))
+	if err != nil {
+		return nil, err
+	}
+	fi, err := f.Stat()
+	if err != nil {
+		f.Close()
+		return nil, err
+	}
+	return &fileImageReader{File: f, size: fi.Size()}, nil
+}
+
+// ImageLoadParams provides one or more values for an external parameter that
+// is supplied to an image which is loaded during the boot process.
+type ImageLoadParams interface {
+	apply(params ...loadParams) []loadParams
```

Comment:
given how it seems to work, applyTo might be a better name for this method?

Reply:
I've renamed this.